### PR TITLE
fix: retry when overdraft peer might be closest

### PR DESCRIPTION
**Pull request by notanatol**
### Checklist

- [X] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)

### Description
[Push Sync closestPeer overdraft case](https://hackmd.io/oZTAHIZjRdm6PUcimWY20Q)

Closes: #3048 

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3052)

### CODE REVIEW

**Review by istae — APPROVED** [review]

File: pkg/pushsync/pushsync.go [hunk 1]
```diff
@@ -375,8 +377,15 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
 					return swarm.ZeroAddress, ErrOutOfDepthStoring
 				}
 
-				ps.pushToNeighbourhood(ctx, fullSkipList, ch, origin, originAddr)
-				return swarm.ZeroAddress, err
+				if skipPeers.OverdraftListEmpty() {
```
[the remark this answers is not in the document]
Comment (notanatol) [reply pkg/pushsync/pushsync.go]: added

File: pkg/pushsync/pushsync.go [hunk 2]
```diff
@@ -397,19 +406,23 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
...
 			go func() {
 				ctxd, cancel := context.WithTimeout(ctx, defaultTTL)
 				defer cancel()
 				ps.pushPeer(ctxd, resultChan, doneChan, peer, ch, origin)
 			}()
 
+		CONT:
 			// reached the limit, do not set timer to retry
 			if allowedRetries <= 0 || allowedPushes <= 0 {
 				continue
```
Comment (aloknerurkar) [inline pkg/pushsync/pushsync.go:428]: If we continue here after an errContinue, the select case would just hang. This assumes a push has been initiated and we wait for the resultChan to unblock us.

**Review by aloknerurkar — CHANGES_REQUESTED** [review]

**Review by notanatol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3052)* status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @acud, @aloknerurkar, @istae, @metacertain, @mrekucci, and @notanatol)

___
*[`pkg/skippeers/skip_peers.go` line 36 at r2](https://reviewable.io/reviews/ethersphere/bee/3052#-N5nuj_HA5yfwVDCswJh:-N5oRI2-5u61mpqQaQgz:b-d6t1pw) ([raw file](https://github.com/ethersphere/bee/blob/8634f24e2ac253f38dc2f3234c73f9427666de10/pkg/skippeers/skip_peers.go#L36)):*
<details><summary><i>Previously, mrekucci wrote…</i></summary><blockquote>

I'd suggest using `map` in order to avoid iterating the whole set before adding a new element.
</blockquote></details>

for small datasets a slice is cheaper than a map, if I'm not mistaken.

**Review by mrekucci — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3052)* status: all files reviewed, 3 unresolved discussions (waiting on @acud, @aloknerurkar, @istae, @metacertain, and @notanatol)

___
*[`pkg/skippeers/skip_peers.go` line 36 at r2](https://reviewable.io/reviews/ethersphere/bee/3052#-N5nuj_HA5yfwVDCswJh:-N5oncaW7SA_TUtf4XeK:b-lz8pvc) ([raw file](https://github.com/ethersphere/bee/blob/8634f24e2ac253f38dc2f3234c73f9427666de10/pkg/skippeers/skip_peers.go#L36)):*
<details><summary><i>Previously, notanatol (Anatol) wrote…</i></summary><blockquote>

for small datasets a slice is cheaper than a map, if I'm not mistaken.
</blockquote></details>

It's up to a ~100 entries.

File: pkg/pushsync/pushsync.go [hunk 3]
```diff
@@ -463,6 +485,8 @@ func (ps *PushSync) measurePushPeer(t time.Time, err error, origin bool) {
 	ps.metrics.PushToPeerTime.WithLabelValues(status).Observe(time.Since(t).Seconds())
 }
 
+var ErrNotAttempted = errors.New("peer not attempted")
```
Comment (istae) [inline pkg/pushsync/pushsync.go:488]: no need to export, and can be moved to the top of the protocol

File: pkg/pushsync/pushsync.go [hunk 4]
```diff
@@ -345,8 +346,9 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
 	}
 
 	var (
-		includeSelf = ps.isFullNode
-		skipPeers   []swarm.Address
+		includeSelf          = ps.isFullNode
+		skipPeers            = new(skippeers.List)
+		errOverdraftNotEmpty = errors.New("continue")
```
Comment (istae) [inline pkg/pushsync/pushsync.go:351]: replace with a better error message

File: pkg/pushsync/pushsync.go [hunk 5]
```diff
@@ -393,16 +402,25 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
...
 			// decrement here to limit inflight requests, if the request is not "attempted", we will increment below
-			allowedPushes--
 
-			peer, err := nextPeer()
+			peer, retry, err := nextPeer()
 			if err != nil {
 				return nil, err
 			}
 
+			if retry {
```
Comment (istae) [inline pkg/pushsync/pushsync.go:411]: much cleaner, thanks :)

**Review by istae — CHANGES_REQUESTED** [review]

**Review by istae — APPROVED** [review]

**Review by aloknerurkar — APPROVED** [review]

**Review by metacertain — APPROVED** [review]

**Review by notanatol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3052)* status: 3 of 4 files reviewed, 8 unresolved discussions (waiting on @acud, @aloknerurkar, @istae, @metacertain, @mrekucci, and @notanatol)

___
*[`pkg/pushsync/pushsync.go` line 415 at r5](https://reviewable.io/reviews/ethersphere/bee/3052#-N5iR_rv5624rb4Qn9DZ-r5-415:-N67g98A8NFgXjN7oFek:bb7f2mf) ([raw file](https://github.com/ethersphere/bee/blob/b1358060cf362a93debdce19d209cd5cf672dd9c/pkg/pushsync/pushsync.go#L415)):*
<details><summary><i>Previously, istae (istae) wrote…</i></summary><blockquote>

can be replaced with a variable, @metacertain is half a second sufficient?
</blockquote></details>

done

File: pkg/skippeers/skip_peers.go [hunk 6]
```diff
@@ -2,38 +2,34 @@
...
-func newSkipPeers() *skipPeers {
-	return &skipPeers{}
 }
 
-func (s *skipPeers) All() []swarm.Address {
+func (s *List) All() []swarm.Address {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	return append(append(s.addresses[:0:0], s.addresses...), s.overdraftAddresses...)
```
[the remark this answers is not in the document]
Comment (notanatol) [reply pkg/skippeers/skip_peers.go]: fixed

**Review by notanatol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3052)* status: 1 of 13 files reviewed, 8 unresolved discussions (waiting on @acud, @aloknerurkar, @istae, @metacertain, and @mrekucci)

___
*[`pkg/pushsync/pushsync.go` line 411 at r5](https://reviewable.io/reviews/ethersphere/bee/3052#-N5iR_rv5624rb4Qn9DZ-r5-411:-N6CL_571Ig-YGx_QILI:b-8fx2u7) ([raw file](https://github.com/ethersphere/bee/blob/b1358060cf362a93debdce19d209cd5cf672dd9c/pkg/pushsync/pushsync.go#L411)):*
<details><summary><i>Previously, istae (istae) wrote…</i></summary><blockquote>

much cleaner, thanks :)
</blockquote></details>

:+1:

**Review by notanatol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3052)* status: 2 of 4 files reviewed, 9 unresolved discussions (waiting on @acud, @aloknerurkar, @istae, @metacertain, and @mrekucci)

___
*[`pkg/pushsync/pushsync.go` line 361 at r7](https://reviewable.io/reviews/ethersphere/bee/3052#-N6D3j3f8MG339gbJ3GB:-N6DcIgq5R8oQl7qOnHv:by7762h) ([raw file](https://github.com/ethersphere/bee/blob/a727d6cc393f121783712b8feaea17398825fcd8/pkg/pushsync/pushsync.go#L361)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

it is not clear what this boolean value means. it would be nice to have named returns here at least for the sake of clarity, or at least adequate comments describing what is being returned
</blockquote></details>

added comment and names
___
*[`pkg/skippeers/skip_peers.go` line 23 at r4](https://reviewable.io/reviews/ethersphere/bee/3052#-N5iycgzIMgZS924T-47-r4-23:-N6DcA_y6RzxUQwagrwU:b6o6g7f) ([raw file](https://github.com/ethersphere/bee/blob/16c7872296331ac27d24e596713ae871c146b050/pkg/skippeers/skip_peers.go#L23)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

Maybe it would be good to have a unit test, to see that returning the appended output does not result in a grow of the field (i.e. repeated calls with different output return the expected output, not resulting in a grow of the slice, consider also mutating the appended content to see that the returned value does not mutate).
</blockquote></details>

I replaced it with an explicit copy-on-read version, pls review

**Pull request merged**